Truncate integer types smaller than 64-bit

lucasccordeiro opened the pull request:
This PR should fix https://github.com/esbmc/esbmc/issues/578.

lucasccordeiro committed:
[simplifier] Truncate integer types < 64-bit

Truncate integer types that are smaller than 64-bit. In order
to achieve that, we need to check the operand bit-width and
then perform the required typecast.

lucasccordeiro committed:
[regression] Added regression tests for bit-shift

Added regression tests for bit-shift operations.
In these regression tests, integer types smaller than 64-bit
should be truncated.

rafaelsamenezes reviewed (approved):
LGTM

mikhailramalho commented on src/util/simplify_expr2.cpp:
```diff
@@ -1289,7 +1289,25 @@ static expr2tc do_bit_munge_operation(
+    default:
+      abort();
+    }
+  }
```
shouldn't we do that for the signedbv case as well?

lucasccordeiro replied:
Fixed.

mikhailramalho commented:
btw, this is just an example of having a BigInt class that doesn't support all operations, so we have to convert back and forth from int64.

We should kill this BigInt eventually and use something more flexible. Maybe boost already has something better to offer.

fbrausse commented on regression/esbmc/github_578_success2/main.c:
```diff
@@ -0,0 +1,10 @@
+int main()
+{
+  int32_t x = 2;
+  x <<= 31;
```
Is this really defined behaviour? I was under the impression that left-shifts touching "the sign bit" for any signed integer type are undefined.

fbrausse replied:
I checked via `-fsanitize=undefined`:
```
../regression/esbmc/github_578_success2/main.c:6:5: runtime error: left shift of 2 by 31 places cannot be represented in type 'int32_t' (aka 'int')
```
The verdict should be `VERIFICATION FAILED`.

fbrausse reviewed (commented):
Maybe it is worth being more generic here based on the `type->get_width()`, e.g. with
```c
static BigInt ones(unsigned n_bits)
{
  BigInt r;
  r.setPower2(n_bits);
  return r -= 1;
}
```
I'm of course talking with my CHERI-hat on, which would benefit from the `int128_t` support, c.f., #404.

fbrausse commented:
> We should kill this BigInt eventually and use something more flexible. Maybe boost already has something better to offer.

I've seen #181, but based on my previous experience it might also be worth considering plain [GMP](https://gmplib.org/manual/C_002b_002b-Interface-General) and/or [Flint](http://www.flintlib.org/) (the latter of which I am using in the [`ksmt`](http://informatik.uni-trier.de/~brausse/ksmt/index.xhtml) solver).

mikhailramalho commented:
> > We should kill this BigInt eventually and use something more flexible. Maybe boost already has something better to offer.
> 
> I've seen #181, but based on my previous experience it might also be worth considering plain [GMP](https://gmplib.org/manual/C_002b_002b-Interface-General) and/or [Flint](http://www.flintlib.org/) (the latter of which I am using in the [`ksmt`](http://informatik.uni-trier.de/~brausse/ksmt/index.xhtml) solver).

tbh, I'm against gmp because of the dreadful API :'(

There's also the problem regarding GPL, as it may conflict with our license.

fbrausse commented:
> tbh, I'm against gmp because of the dreadful API :'(

Yeah... I'm using either of them via the [kay](https://github.com/fbrausse/kay) wrapper. My main problem with the gmp API is that they put it into global namespace.

> There's also the problem regarding GPL, as it may conflict with our license.

Does the LPGL (either v2.1+ for flint or v3+ for gmp) also conflict with the current license?

mikhailramalho commented:
Em seg., 3 de jan. de 2022 às 16:04, Franz Brauße ***@***.***>
escreveu:

> tbh, I'm against gmp because of the dreadful API :'(
>
> Yeah... I'm using either of them via the kay
> <https://github.com/fbrausse/kay> wrapper. My main problem with the gmp
> API is that they put it into global namespace.
>
> There's also the problem regarding GPL, as it may conflict with our
> license.
>
> Does the LPGL (either v2.1+ for flint or v3+ for gmp) also conflict with
> the current license?
>

I don't remember the details but Jeremy checked that a few years ago and
there were some issues with GPL/LGPL. I remember even suggesting a shim
layer but that was too much of a gray area for us to proceed.

—
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/579#issuecomment-1004293589>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH5RV6WWCTREQ3HWYX3UUHXLVANCNFSM5LFVVAFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>


-- 

Mikhail Ramalho.

fbrausse commented:
I've added a test-case which still fails without `--no-simplify`. Could you take a look?

lucasccordeiro commented:
> I've added a test-case which still fails without --no-simplify. Could you take a look?

Okay. I got it.

mikhailramalho commented on src/util/simplify_expr2.cpp:
```diff
@@ -1287,9 +1287,51 @@ static expr2tc do_bit_munge_operation(
+    default:
+      abort();
+    }
+  }
```
That's a lot of duplicated code... Can we make this code more generic? 

Something like:
```
uint64_t bar = static_cast<uint64_t>(val1) & ((1ULL << type->get_width()) - 1);
return constant_int2tc(type, BigInt(bar));
```

I think it should cover both signed and unsigned cases, otherwise we can keep the `if(is_signedbv_type(type))` branch.

lucasccordeiro replied:
I agree with you, @mikhailramalho. 

I have rewritten the code after a quick meeting with @fbrausse. This suggestion seems to be a clean solution; let us see if the CI will not complain about that.

mikhailramalho committed:
We don't care about signedness

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho commented:
So, the reason my suggestions didn't work was because (1ULL << 64) is undef behaviour :)

I pushed a small change, @lucasccordeiro, to remove the `if(is_signedbv_type(type))` and got no regressions locally, let's see our CI. Feel free to remove/change my commit if needed.

fbrausse commented:
The failing test `cbmc/01_cbmc_BV_Arithmetic4` uses undefined behaviour by left-shifting an `int` until all bits are shifted out, so technically `VERIFICATION FAILED` is the correct result. `-fsanitize=undefined` output of the test:
```
../regression/cbmc/01_cbmc_BV_Arithmetic4/main.c:10:10: runtime error: left shift of 1073741824 by 1 places cannot be represented in type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../regression/cbmc/01_cbmc_BV_Arithmetic4/main.c:10:10 in 
../regression/cbmc/01_cbmc_BV_Arithmetic4/main.c:11:9: runtime error: left shift of 1073741824 by 1 places cannot be represented in type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../regression/cbmc/01_cbmc_BV_Arithmetic4/main.c:11:9 in 
```
Fixing the test case so it does not produce UB also results in the correct ESBMC verdict:
```diff
diff --git a/regression/cbmc/01_cbmc_BV_Arithmetic4/main.c b/regression/cbmc/01_cbmc_BV_Arithmetic4/main.c
index 3321bcb90d..e49137ada6 100644
--- a/regression/cbmc/01_cbmc_BV_Arithmetic4/main.c
+++ b/regression/cbmc/01_cbmc_BV_Arithmetic4/main.c
@@ -1,8 +1,8 @@
 /* Copy new sign bit to the left */
 int test_extend2(int x, unsigned width)
 {
-  int mask = 1 << (width-1);
-  int bit = (x & mask) << 1;
+  unsigned mask = 1 << (width-1);
+  unsigned bit = (x & mask) << 1;
   mask <<= 1;
   while (mask) {
     x = bit | (~mask & x);
```
For reference, on the original test case, esbmc believes `r == -1 << 2` and `-4 == -1 << 2` to be true but `r == -4` not. This might be a different issue.

lucasccordeiro committed:
[Regression] Updated verdict for bitshit case

Updated verdict of a regression test that uses undefined
behaviour by left-shifting an int until all bits are
shifted out.

lucasccordeiro commented:
> The failing test cbmc/01_cbmc_BV_Arithmetic4 uses undefined behaviour by left-shifting an int until all bits are shifted out, so technically VERIFICATION FAILED is the correct result.

@fbrausse: many thanks for this assessment of the test case. I have just updated the PR accordingly to consider your analysis.

fbrausse reviewed (changes requested):
`regression/esbmc/github_578_success2/main.c` should be `VERIFICATION FAILED` since it indeed exhibits undefined behaviour:
```
../regression/esbmc/github_578_success2/main.c:6:5: runtime error: left shift of 2 by 31 places cannot be represented in type 'int32_t' (aka 'int')
```

fbrausse committed:
[util/simplify_expr2] simplify do_bit_munge_operation()

lucasccordeiro commented:
> regression/esbmc/github_578_success2/main.c should be VERIFICATION FAILED since it indeed exhibits undefined behaviour:

Fixed.

mikhailramalho commented:
> The failing test `cbmc/01_cbmc_BV_Arithmetic4` uses undefined behaviour by left-shifting an `int` until all bits are shifted out, so technically `VERIFICATION FAILED` is the correct result. `-fsanitize=undefined` output of the test:
> 
> ```
> ../regression/cbmc/01_cbmc_BV_Arithmetic4/main.c:10:10: runtime error: left shift of 1073741824 by 1 places cannot be represented in type 'int'
> SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../regression/cbmc/01_cbmc_BV_Arithmetic4/main.c:10:10 in 
> ../regression/cbmc/01_cbmc_BV_Arithmetic4/main.c:11:9: runtime error: left shift of 1073741824 by 1 places cannot be represented in type 'int'
> SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../regression/cbmc/01_cbmc_BV_Arithmetic4/main.c:11:9 in 
> ```
> 
> Fixing the test case so it does not produce UB also results in the correct ESBMC verdict:
> 
> ```diff
> diff --git a/regression/cbmc/01_cbmc_BV_Arithmetic4/main.c b/regression/cbmc/01_cbmc_BV_Arithmetic4/main.c
> index 3321bcb90d..e49137ada6 100644
> --- a/regression/cbmc/01_cbmc_BV_Arithmetic4/main.c
> +++ b/regression/cbmc/01_cbmc_BV_Arithmetic4/main.c
> @@ -1,8 +1,8 @@
>  /* Copy new sign bit to the left */
>  int test_extend2(int x, unsigned width)
>  {
> -  int mask = 1 << (width-1);
> -  int bit = (x & mask) << 1;
> +  unsigned mask = 1 << (width-1);
> +  unsigned bit = (x & mask) << 1;
>    mask <<= 1;
>    while (mask) {
>      x = bit | (~mask & x);
> ```
> 
> For reference, on the original test case, esbmc believes `r == -1 << 2` and `-4 == -1 << 2` to be true but `r == -4` not. This might be a different issue.

Ouch.

We could easily add VCs to catch such cases, something similar to the --overflow-check.

lucasccordeiro commented:
@mikhailramalho:

> We could easily add VCs to catch such cases, something similar to the --overflow-check.

In this PR, ESBMC correctly reports `Verification Failed` for this specific regression test, as reported by @fbrausse.

mikhailramalho commented:
Yeah Yeah, what I meant is that ESBMC could have a check for that,
something --undef-behaviour-check, which would add an assert(const >= 0)
for shift operations, or assert(const <= type-bit-width).

It should be quite easy to implement similar to the --overflow-check (in
the goto_check file), but it's obviously not to be done in this PR.

Em qua., 5 de jan. de 2022 às 11:31, Lucas Cordeiro <
***@***.***> escreveu:

> @mikhailramalho <https://github.com/mikhailramalho>:
>
> We could easily add VCs to catch such cases, something similar to the
> --overflow-check.
>
> In this PR, ESBMC correctly reports Verification Failed for this specific
> regression test, as reported by @fbrausse <https://github.com/fbrausse>.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/579#issuecomment-1005735542>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH2ZQAVSPLRSKMKX5SLUURI5NANCNFSM5LFVVAFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Mikhail Ramalho.

lucasccordeiro commented:
Oh I see. I have just created this issue https://github.com/esbmc/esbmc/issues/580 to record this suggestion.

fbrausse commented:
After discussing with Lucas, I've added a simplification, which does restore the previous behaviour on the UB test case `cbmc/01_cbmc_BV_Arithmetic4` (and also making it `VERIFICATION SUCCESSFUL` again - since the new UB-check is tbd).

mikhailramalho commented:
Wait, but the test should fail, right?

Why do we want to keep a test case with an incorrect verdict? Regardless of
having a UB check or not.

Em qua., 5 de jan. de 2022 às 13:16, Franz Brauße ***@***.***>
escreveu:

> After discussing with Lucas, I've added a simplification, which does
> restore the previous behaviour on the UB test case
> cbmc/01_cbmc_BV_Arithmetic4 (and also making it VERIFICATION SUCCESSFUL
> again - since the new UB-check is tbd).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/579#issuecomment-1005867793>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH7PDR74WIYU3WTA43TUURVFRANCNFSM5LFVVAFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Mikhail Ramalho.

fbrausse commented:
Yes, it should fail when undefined behaviour is considered. With `--overflow-check` it does. I'll add a copy with this flag. Maybe we should mark the existing one as `KNOWNBUG`?

lucasccordeiro commented:
We could change the verdict and mark it as a known bug since we don't have an option to check for undefined behavior yet.

I'm okay with both ways.

fbrausse committed:
[regression] add copy of cbmc/01_cbmc_BV_Arithmetic4 with --overflow-check set

mikhailramalho commented:
I agree, marking it with knownbug is the way to go.

Em qua., 5 de jan. de 2022 às 13:26, Lucas Cordeiro <
***@***.***> escreveu:

> We could change the verdict and mark it as a known bug since we don't have
> an option to check for undefined behavior yet.
>
> I'm okay with both ways.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/579#issuecomment-1005876317>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH4QFFOCUZZ3E6B5FDLUURWKHANCNFSM5LFVVAFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Mikhail Ramalho.

fbrausse committed:
[regression] mark cbmc/01_cbmc_BV_Arithmetic4 as known bug, see #580

This test case left-shifts a (signed) int until it becomes zero. This is
undefined behaviour.

fbrausse commented:
> mark it as a known bug

Done.
